### PR TITLE
Modularize sandbox support

### PR DESCRIPTION
**Pull request by knightsc**
This change makes it much easier to add support for more sandboxes. Rather than having to continue to expand the code in post_b64.py now a new sandbox module can simply be added with the appropriate configuration. If the sandbox is enabled in the configuration then the post_b64 module will send the binaries it finds to it

### CODE REVIEW

File: pastehunter.py [hunk 1]
```diff
@@ -155,7 +155,7 @@ def paste_scanner():
             sleep(0.5)
         else:
             paste_data = q.get()
-            with timeout(seconds=5):
+            with timeout(seconds=conf['general']['process_timeout']):
```
Comment (Plazmaz) [inline pastehunter.py]: Really nice little addition here.

File: docs/sandboxes.rst [hunk 2]
```diff
@@ -0,0 +1,25 @@
+Sandboxes
```
Comment (Plazmaz) [inline docs/sandboxes.rst]: We should also remove sandboxes from `postprocess.rst` and probably add a mention to the main readme.

File: pastehunter.py [hunk 3]
```diff
@@ -230,6 +230,17 @@ def paste_scanner():
                     # Else use the rule name
                     else:
                         results.append(match.rule)
+
+                # Store additional fields for passing on to post processing
+                encoded_paste_data = raw_paste_data.encode('utf-8')
```
Comment (Plazmaz) [inline pastehunter.py]: Interesting. Why did you move this logic to be before the blacklist check/postprocessing? It shouldn't impact the shas if that was the intent here.

File: postprocess/post_b64.py [hunk 4]
```diff
@@ -55,47 +55,18 @@ def run(results, raw_paste_data, paste_object):
                     # We are guessing that the sample has been submitted, and crafting a URL
                     paste_object["VT"] = 'https://www.virustotal.com/#/file/{0}'.format(paste_object["exe_md5"])
 
-                    # Cuckoo
-                    if conf["post_process"]["post_b64"]["cuckoo"]["enabled"]:
```
Comment (Plazmaz) [inline postprocess/post_b64.py]: One thing I am really concerned about is that this change will break existing configs. I'm wondering if there's any way we could make this change more backwards-compatible. I'm on the fence about if it's worth the technical debt to just say if a postprocessing module matches one of the sandboxes, print a warning and goto sandbox logic.

File: sandboxes/cuckoo.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    except Exception as e:
+        logger.error("Unabled to submit sample to cuckoo")
+
+    # Send any updated json back
+    return paste_object
+
+def send_to_cuckoo(raw_exe, pasteid):
+    cuckoo_ip = conf["sandboxes"]["cuckoo"]["api_host"]
+    cuckoo_port = conf["sandboxes"]["cuckoo"]["api_port"]
+    cuckoo_host = 'http://{0}:{1}'.format(cuckoo_ip, cuckoo_port)
```
Comment (Plazmaz) [inline sandboxes/cuckoo.py]: Can we make this configurable and default to TLS? (assuming cuckoo isn't http-only)

File: sandboxes/viper.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import requests
+from common import parse_config
+conf = parse_config()
+
+logger = logging.getLogger('pastehunter')
+
+def upload_file(raw_file, paste_object):
+    viper_ip = conf["sandboxes"]["viper"]["api_host"]
+    viper_port = conf["sandboxes"]["viper"]["api_port"]
+    viper_host = 'http://{0}:{1}'.format(viper_ip, viper_port)
```
Comment (Plazmaz) [inline sandboxes/viper.py]: Same here, would like http/https to be configurable

File: pastehunter.py (hunk 3 above)
Comment (knightsc) [reply pastehunter.py]: In my case it was specifically because I wanted the rules that matched (`YaraRule`) to be available to the sandbox modules. It seemed to make the most sense to just move this entire block up before postprocessing so that all the final `paste_data` would be available to any post processing module.

File: docs/sandboxes.rst (hunk 2 above)
Comment (knightsc) [reply docs/sandboxes.rst]: Updated in 1382c15

File: postprocess/post_b64.py (hunk 4 above)
Comment (knightsc) [reply postprocess/post_b64.py]: I think my question would be how many users do you think would be affected by the change in config? With the change I don't think it will throw an exception on startup if you're using the old config but it would act as if no sandboxes were enabled. @kevthehermit thoughts?

File: sandboxes/cuckoo.py (hunk 5 above)
Comment (knightsc) [reply sandboxes/cuckoo.py]: Before I make this change do you know why it's http in the first place? Is that the default and normal settings for Cukcoo out of the box. When I approached the change code wise my goal was not to change any of the existing behavior of the sandboxes to make it easier to tell that the refactor didn't break anything. Thus I didn't change anything here logic wise really.

If we do want to make a change would it make more sense to just make a `cuckoo_url` config parameter rather than a tls true false? Something like that would also affect backward compatibility with config stuff though.

File: sandboxes/viper.py (hunk 6 above)
Comment (knightsc) [reply sandboxes/viper.py]: Same comment as https://github.com/kevthehermit/PasteHunter/pull/87#discussion_r329375900

File: sandboxes/cuckoo.py (hunk 5 above)
Comment (Plazmaz) [reply sandboxes/cuckoo.py]: Hmmm, yeah, interesting question. That I'm also not sure about. Will wait for comments from @kevthehermit on this and the other compatibility Q. (If we don't get a response in a few days I think we should be fine to leave the PR as-is)

**Comment by kevthehermit** [comment]
A lot to unpack in there :)

In terms of splitting the sandbox items out so its not dependant on the base64 decoder makes sense, especially as that would provide more flexibility for example passing powershell scripts. 

Theses sandboxes are typically local instances not remote so HTTP is the norm. If its behind HTTPS / TLS its typically via a proxy like nginx or apache. 

For backward compatibility this has been an issue historically and I dont expect that to go away any time soon. Switching to more frequent releases with a release package and versioning will help. Add some kind of warning message on start.

Also have been looking at making this more distributable, deb/ppa / pip installable

Not sure if that helps or hinders this specific thread

**Comment by kevthehermit** [comment]
HIjacking this a little. As there have been a lot of changes; thanks @Plazmaz , Lets create a new release of the back of this PR update changelogs and then figure out a release methodolgy

**Comment by kevthehermit** [comment]
Can you resolve the conflicts, then i can test, merge and create a new release on the back of this one. 
Thanks

**Comment by knightsc** [comment]
@kevthehermit I rebased on the latest code from master and fixed the conflicts
